api.views: add try push handover functionality (bug 2002566)

zzzeid opened the pull request:
- add functionality that adds try_task_config.json file
- update landing worker to process handover jobs
- add PullRequestTryLintAPIView
- add LandingJob.handover_repo field and migrations
- add LandingJob.is_handed_over field and migrations
- add Repo.is_try helper method

zzzeid commented on src/lando/api/legacy/workers/landing_worker.py:
```diff
@@ -215,10 +229,31 @@ def apply_patch(revision: Revision):
+            job.target_repo = job.handover_repo
+            job.handover_repo = None
+            job.save()
+            job.revisions.update(pull_number=None)
```
Add comment describing what is happening here: after the patches are generated, the job is converted from a pull-request "git" landing job to a legacy hg landing job targeting the try repo.

shtrom commented on src/lando/api/views.py:
```diff
@@ -303,3 +303,43 @@ def get(self, request: WSGIRequest, repo_name: str, number: int) -> JsonResponse
+            "author_name": author_name,
+            "author_email": author_email,
+            "commit_message": pull_request.commit_message,
+            "timestamp": int(datetime.now().timestamp()),
```
You could set the timestamp to the pull requests's `updatedAt` field. As far as I have seen, it gets updaded reliably with any comments and other metadata changes, in addition to plain code changes.

zzzeid commented on src/lando/api/legacy/workers/landing_worker.py:
```diff
@@ -168,6 +169,19 @@ def run_job(self, job: LandingJob) -> bool:
+                {
+                    "parameters": {
+                        "optimize_target_tasks": True,
+                        "target_tasks_method": "codereview",
```
>These informations should be provided by the endpoint caller, so I guess you only lack the head commit in the URL or payload.

The "repo name" in the URL is actually a little different than what the GitHub repo name is (at least in the firefox repo case). The repo name we use in Lando is a combination of the repo name + branch name, e.g., `firefox-main`. This would be the portion that is provided in the URL. An alternative, if really needed, can be to provide a repo + branch name and have Lando look up the "Lando repo" based on that info.

The GitHub repo name and branch are both also likely needed in the payload, regardless of the URL format.

EDIT: an alternative would also be to use the target branch of the pull request (i.e., "base branch") to determine the Lando repo, however, that would be more complex on the Lando side as we have to authenticate with a repo before we query for pull requests.

zzzeid commented on src/lando/main/models/repo.py:
```diff
@@ -101,6 +101,10 @@ class HooksChoices(models.TextChoices):
 
+    @property
+    def is_try(self) -> bool:
+        return self.name == "try"
```
I think this is possible in the future when we migrate try to a git repo, however currently this mimics the existing try repo already set up in legacy Lando (i.e., hg.mozilla.org/try). The repo name is technically `try` and I think until we move it to a different repo. Thoughts?

shtrom commented on src/lando/main/models/repo.py:
```diff
@@ -101,6 +101,10 @@ class HooksChoices(models.TextChoices):
         return str(self.system_path) or self.get_system_path()
 
+    @property
+    def is_try(self) -> bool:
```
Yep, ok. I see the difference now.

I think we should do the renaming, though, but I otherwise agree with the rest.

shtrom commented on src/lando/api/legacy/workers/landing_worker.py:
```diff
@@ -215,10 +229,31 @@ def apply_patch(revision: Revision):
 
+        if job.is_pull_request_job and job.handover_repo and job.handover_repo.is_try:
+            self.add_try_task_config(scm)
+            self.convert_patches_to_diff(scm, job)
```
Something like an `addremove`? We use `add -A` in `get_diff_from_patches` https://github.com/mozilla-conduit/lando/blob/425adeac414da821075e4b22f94a0d3ebb999e64/src/lando/main/scm/git.py#L220 Is that not sufficient?

zzzeid replied:
Renamed to `add_try_task_config` so that this is more explicit.

shtrom commented on src/lando/api/legacy/workers/landing_worker.py:
```diff
@@ -230,6 +244,20 @@ def apply_patch(revision: Revision):
 
         self.update_repo(repo, job, scm, job.target_commit_hash)
 
+        if job.is_pull_request_job and job.handover_repo and job.handover_repo.is_try:
```
I'd suggest moving the `is_try` down, as a failing case, so we don't continue over this check if something is inconsistent
```suggestion
        if job.is_pull_request_job and job.handover_repo:
            if not job.handover_repo.is_try:
                raise Exception("Cannot hand-over to non-try repos.")
```

shtrom commented on src/lando/api/views.py:
```diff
@@ -305,3 +305,43 @@ def get(self, request: WSGIRequest, repo_name: str, number: int) -> JsonResponse
+            is_handed_over=False,
+            handover_repo=try_repo,
+            requester_email=ldap_username,
+        )
```
Do we need to specify `is_pull_request_job=True,` here?
```suggestion
        job = LandingJob.objects.create(
            target_repo=target_repo,
            is_handed_over=False,
            handover_repo=try_repo,
            requester_email=ldap_username,
            is_pull_request_job=True,
        )
```

shtrom commented on src/lando/main/models/landing_job.py:
```diff
@@ -48,6 +48,18 @@ class LandingJob(BaseJob):
+        blank=True,
+        related_name="handover_jobs",
+    )
+    is_handed_over = models.BooleanField(null=True, blank=True)
```
How about

```suggestion
    is_handed_over = models.BooleanField(default=False)
```

shtrom reviewed (commented):
A couple of small nits, but this is looking good to me.

shtrom reviewed (approved):
lgtm. As a nit, I'd rather not change the name of `scm.get_diff_from_patches`, but we're good to go in otherwise.

zzzeid committed:
api.views: add try push handover functionality (bug 2002566)

- add functionality that adds try_task_config.json file
- update landing worker to process handover jobs
- add PullRequestTryLintAPIView
- add LandingJob.handover_repo field and migrations
- add LandingJob.is_handed_over field and migrations
- add Repo.is_try helper method
